[15.0][IMP] purchase_security: Add groups to purchase orders and new group …

stefan-tecnativa opened the pull request:
…permission feature

cc @Tecnativa TT44023

@pedrobaeza @victoralmau please review!

What has been done is to add groups to purchase orders, along with a new "purchase group" permission. So users who have this permission and belong to the group associated with the purchase order, will be able to see the purchase orders of other users with whom they share that group.

stefan-tecnativa committed:
[IMP] purchase_security: Add groups to purchase orders and new group permission feature

TT44023

pedrobaeza commented on purchase_security/models/purchase.py:
```diff
@@ -0,0 +1,13 @@
+    _inherit = "purchase.order"
+
+    purchase_group_ids = fields.Many2one(
+        "res.groups",
```
This should be a new model: `purchase.team`. And call this field `team_id` (without `s`), as it's a many2one and it's inside `purchase.order`, so `purchase` is redundant.

pedrobaeza commented on purchase_security/security/security.xml:
```diff
@@ -47,5 +55,13 @@
             <field name="groups" eval="[(4,ref('group_purchase_own_orders'))]" />
         </record>
+        <record model="ir.rule" id="purchase_order_user_group_purchases_rule">
+            <field name="name">View purchase orders (Group purchases)</field>
```
Follow the nomenclature of the rest of the groups:

![Selección_037](https://github.com/OCA/purchase-workflow/assets/7165771/acde18a2-fa19-438a-a7e7-add4198ebfd8)


```suggestion
            <field name="name">User (team orders)</field>
```

And even more, you have broken the inheritance, as you have to change the "Administrator" to inherit from this new group.

stefan-tecnativa replied:
Thanks for all the suggestions @pedrobaeza , now I have a better understanding of the approach to be taken. I made some modifications, and I am aware that i have not added yet some view for the new model. Review my changes when you have time please.

pedrobaeza reviewed (changes requested):
You need to add as well a `purchase.order` rule for the new group.

Check `sales_team` for mimicking it on purchases.

pedrobaeza commented on purchase_security/models/purchase.py:
```diff
@@ -0,0 +1,18 @@
+    _description = "Purchase Team"
+
+    name = fields.Char()
+    user_id = fields.Many2many("res.users", string="Purchase Users")
```
Please respect conventions. m2m/o2m always with `_ids`.

```suggestion
    user_ids = fields.Many2many("res.users", string="Purchase Users")
```

pedrobaeza commented on purchase_security/models/purchase.py:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2023 Tecnativa - Stefan Ungureanu
```
Only one model per file, unless it's a "children" model, so this file should be called `purchase_team.py`, and only contains the first class. The other should be put in a file `purchase_order.py`. If you also inherit `purchase.order.line`, then that one can fit inside `purchase_order.py`.

pedrobaeza commented on purchase_security/models/res_users.py:
```diff
@@ -0,0 +1,10 @@
+class ResUsers(models.Model):
+    _inherit = "res.users"
+
+    purchase_team_id = fields.Many2many("purchase.team", string="Purchase Teams")
```
No need of this inverse field.

pedrobaeza commented on purchase_security/security/ir.model.access.csv:
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+purchase_security.access_purchase_team,access_purchase_team,purchase_security.model_purchase_team,base.group_user,1,1,1,1
```
This permission is too broad. That's a bad practice. If this is for purchase, why everything is able to modify it? Put read permission for purchase users, and full access to purchase admins.

pedrobaeza commented on purchase_security/security/security.xml:
```diff
@@ -30,6 +38,15 @@
+            <field name="model_id" ref="purchase.model_purchase_order" />
+            <field
+                name="domain_force"
+            >['|', '|', ('user_id', '=', user.id), ('user_id', '=', False), '|', ('team_id', '=', user.purchase_team_id.id),('team_id', '=', False)]
```
Why adding the user part? That one is already covered by the inferior record rule, isn't it?

stefan-tecnativa replied:
Thanks for all the comments. You are right, I wasn't fully understanding the difference between this rules, thats wh i was checking the user twice. I made the suggested changes.

pedrobaeza commented on purchase_security/models/purchase.py:
```diff
@@ -0,0 +1,18 @@
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    team_id = fields.Many2one("purchase.team", string="Team")
```
```suggestion
    team_id = fields.Many2one("purchase.team", string="Team", index=True)
```

pedrobaeza reviewed (changes requested):
Missing menu entry and view for purchase teams. The commit message shouldn't be `[WIP] ...`.

chienandalu commented on purchase_security/security/security.xml:
```diff
@@ -30,6 +38,15 @@
             <field name="groups" eval="[(4,ref('group_purchase_own_orders'))]" />
         </record>
+        <record model="ir.rule" id="purchase_order_group_purchase_order_group_orders">
+            <field name="name">View purchase orders (group responsible)</field>
```
```suggestion
            <field name="name">View purchase orders (purchase team member)</field>
```

chienandalu commented on purchase_security/security/security.xml:
```diff
@@ -30,6 +38,15 @@
+            <field name="model_id" ref="purchase.model_purchase_order" />
+            <field
+                name="domain_force"
+            >['|', ('team_id', '=', user.purchase_team_id.id),('team_id', '=', False)]
```
That field `purchase_team_id` on `res.users` doesn't even exists. How is this supposed to work?

stefan-tecnativa replied:
Thanks for the comments, can you check if now it is correct?
I will improve the test coverage after.

chienandalu reviewed (changes requested):
Some comments:

- The module doesn't work :confused: . You can figure out why in the comments below...
- In that regard, I'm missing some tests for these new features.
- Should team members be subscribed automatically to team POs?
- Something with this module is making the user purchase permissions show up as checkboxes instead of a select:

![image](https://github.com/OCA/purchase-workflow/assets/5040182/13586e1a-5c73-4b6c-b9d7-b6a86a8f2b9c)

pedrobaeza commented:
> Should team members be subscribed automatically to team POs?

Out of scope of this PR for now. It can be put in the ROADMAP though.

chienandalu commented on purchase_security/security/security.xml:
```diff
@@ -47,5 +65,13 @@
+            <field name="model_id" ref="purchase.model_purchase_order_line" />
+            <field
+                name="domain_force"
+            >['|',(user.id, 'in', 'order_id.team_id.user_ids'),('order_id.team_id', '=',False)]</field>
```
@stefan-tecnativa please at least make the minimum functional tests before asking for reviews. This is wrong

chienandalu commented on purchase_security/security/security.xml:
```diff
@@ -8,10 +8,19 @@
             <field
                 name="implied_ids"
-                eval="[(3, ref('purchase.group_purchase_user')), (4, ref('group_purchase_own_orders'))]"
+                eval="[(3, ref('purchase.group_purchase_user')),
+                (4, ref('group_purchase_own_orders')),(4, ref('group_purchase_group_orders'))]"
```
I don't think you need to imply these groups as managers override the access rules

chienandalu commented on purchase_security/views/purchase_team_views.xml:
```diff
@@ -0,0 +1,37 @@
+                <form string="Purchase Team">
+                    <group name="team">
+                        <field name="name" />
+                        <field name="user_ids" />
```
Maybe you can customize the tree a little (or use a different one) so we don't have much columns or you can copy from the `sales_team.crm_team_view_form` view that uses a kanban which is more visually pleasant.

chienandalu commented on purchase_security/security/security.xml:
```diff
@@ -47,5 +65,13 @@
+            <field name="model_id" ref="purchase.model_purchase_order_line" />
+            <field
+                name="domain_force"
+            >['|', ('order_id.team_id.user_ids', 'in', [user.id]), ('order_id.team_id', '=', False)]</field>
```
This is not going to be very performant. What do you think @pedrobaeza ?

pedrobaeza replied:
We can put `auto_join=True` for having less penalty, and it's compatible with security.

stefan-tecnativa replied:
Thanks for all the suggestions!

chienandalu reviewed (commented):
Now it seems to work. Some comments:

chienandalu commented on purchase_security/tests/test_access_rights.py:
```diff
@@ -14,15 +14,56 @@ def setUpClass(cls):
+        cls.team1 = cls.env["purchase.team"].create({"name": "Team1"})
+        cls.team2 = cls.env["purchase.team"].create({"name": "Team2"})
+        group_name_own = "group_purchase_own_orders"
+        group_name_team = "group_purchase_group_orders"
```
Don't use string formating for this. It isn't very useful and removes clarity to the case (I know it was like that before, but you can fix it now)

chienandalu commented on purchase_security/models/purchase_team.py:
```diff
@@ -0,0 +1,12 @@
+    _description = "Purchase Team"
+
+    name = fields.Char()
+    user_ids = fields.Many2many("res.users", string="Purchase Users", auto_join=True)
```
Do Many2many have auto_join?

pedrobaeza replied:
No, they are automatically auto-joined in ORM. The auto join was for `team_id` many2one (and not sure right now if with `-`, `_`, or nothing).

stefan-tecnativa replied:
Now i got it. I haven't found any other kind of declaration for this parameter than the one you provided to me.

chienandalu replied:
I case of doubt, a natural step is to go to the docs: https://www.odoo.com/documentation/15.0/developer/reference/backend/orm.html?#relational-fields

pedrobaeza committed:
[IMP] purchase_security: Computed and sortable team

- Fine-tune security rules.
- Put sequence on teams.
- Edit teams on tree.
- Convert purchase order team in computed writable.
- Use new_test_user in tests + useless lines for shorten code.
- Extra tests

pedrobaeza reviewed (approved):
/ocabot merge major
